### PR TITLE
Socks 5 proxy: Only use mgmt cluster for cloud apis for ingress operator

### PR DESCRIPTION
**Pull request by alvaroaleman**
The socks5 proxy has code to avoid sending traffic destined for cloud
provider apis through konnektivity in order to not require nodes to be
up for that communication to work.

When running a management cluster with a proxy setup and not setting
proxy env vars in the socks5 proxy, that code effectively means that it
is not possible to commnicate with those apis, as the sidecar can not
route it as it has no direct Internet connectivity.

The above currently breaks imagestreamimports if the import is from a docker
registry hosted in AWS S3, as the Openshift APIServer is supposed to
route all its imports through Konnektivity, which is why its proxy
sidecar does not get the mgmt cluster proxy env vars.

The only reason for special casing cloud provider apis here is the
ingress operator, as it should be able to setup DNS before nodes exist
in order to speed up cluster creation.

As this special case is surprising and an easy footgun, hide it behind a
flag and only activate this for the Ingress Operator.

This change should stabilize the proxy conformance periodic, which has
currently a high failure rate which ultimately always boilds down to
being unable to import the `tools` image in the `openshift` namespace,
which is used by many tests.

/cc @csrwng 

Likely the longest commit message for any commit authored by me in this repo :upside_down_face: 

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, use `fixes #<issue_number>(, fixes #<issue_number>, ...)` format, where issue_number might be a GitHub issue, or a Jira story*:
Fixes #

**Checklist**
- [ ] Subject and description added to both, commit and PR.
- [ ] Relevant issues have been referenced.
- [ ] This change includes docs. 
- [ ] This change includes unit tests.

### CODE REVIEW

**Comment by alvaroaleman** [comment]
Sample log from https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/periodic-ci-openshift-hypershift-main-periodics-e2e-conformance-proxy/1549997772126556160 showing the openshift apiserver socks5 proxy sidecar being unable to connect to the bucket url: https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/logs/periodic-ci-openshift-hypershift-main-periodics-e2e-conformance-proxy/1549997772126556160/artifacts/e2e-conformance-proxy/dump/artifacts/namespaces/clusters-7328c508-08ba/core/pods/logs/openshift-apiserver-68fb48c474-z87cb-socks5-proxy.log

File: konnectivity-socks5-proxy/main.go [hunk 1]
```diff
@@ -39,10 +39,12 @@ func NewStartCommand() *cobra.Command {
 	var caCertPath string
 	var clientCertPath string
 	var clientKeyPath string
+	var connectDirectlyToCloudApis bool
 
 	cmd.Flags().StringVar(&proxyHostname, "konnectivity-hostname", "konnectivity-server-local", "The hostname of the konnectivity service.")
 	cmd.Flags().IntVar(&proxyPort, "konnectivity-port", 8090, "The konnectivity port that socks5 proxy should connect to.")
 	cmd.Flags().IntVar(&servingPort, "serving-port", 8090, "The port that socks5 proxy should serve on.")
+	cmd.Flags().BoolVar(&connectDirectlyToCloudApis, "connect-directly-to-cloud-apis", false, "If traffic destined for AWS or Azure apis should be sent there directly rather than going through konnectivity. If enabled, proxy env vars from the mgmt cluster must be propagated to this container")
```
Comment (csrwng) [inline konnectivity-socks5-proxy/main.go:47]: naming/grammar nits:

s/connectDirectlyToCloudApis/connectDirectlyToCloudAPIs/
s/If traffic destined/If true, traffic destined/ 
s/AWS or Azure apis/AWS or Azure APIs/

File: konnectivity-socks5-proxy/main.go [hunk 2]
```diff
@@ -74,9 +76,9 @@ func NewStartCommand() *cobra.Command {
 	return cmd
 }
 
-func dialFunc(caCertPath string, clientCertPath string, clientKeyPath string, proxyHostname string, proxyPort int) func(ctx context.Context, network string, addr string) (net.Conn, error) {
+func dialFunc(caCertPath string, clientCertPath string, clientKeyPath string, proxyHostname string, proxyPort int, connectDirectlyToCloudApis bool) func(ctx context.Context, network string, addr string) (net.Conn, error) {
 	return func(ctx context.Context, network string, addr string) (net.Conn, error) {
-		if shouldGoDirect(strings.Split(addr, ":")[0]) {
+		if connectDirectlyToCloudApis && shouldGoDirect(strings.Split(addr, ":")[0]) {
```
Comment (csrwng) [inline konnectivity-socks5-proxy/main.go:81]: `shouldGoDirect` should probably be named something like `isCloudAPI`

**Review by csrwng — COMMENTED** [review]
Awesome stuff. A couple of naming nits.

File: konnectivity-socks5-proxy/main.go (hunk 2 above)
Comment (alvaroaleman) [reply konnectivity-socks5-proxy/main.go]: That is a great idea, updated, ptal

File: konnectivity-socks5-proxy/main.go (hunk 1 above)
Comment (alvaroaleman) [reply konnectivity-socks5-proxy/main.go]: updated

**Pull request merged**